pandas/tests/reshape/merge/test_multi.py

```diff
@@ -0,0 +1,623 @@
+                                                            kind='mergesort')
+        tm.assert_frame_equal(left, right)
+
+    def test_left_join_multi_index(self):
+        icols = ['1st', '2nd', '3rd']
+
+        def bind_cols(df):
+            def iord(a):
+                return 0 if a != a else ord(a)
+
+            def f(ts):
+                return ts.map(iord) - ord('a')
+
+            return (f(df['1st']) + f(df['3rd']) * 1e2 +
+                    df['2nd'].fillna(0) * 1e4)
+
+        def run_asserts(left, right):
+            for sort in [False, True]:
+                res = left.join(right, on=icols, how='left', sort=sort)
+
+                assert len(left) < len(res) + 1
+                assert not res['4th'].isna().any()
+                assert not res['5th'].isna().any()
+
+                tm.assert_series_equal(
+                    res['4th'], - res['5th'], check_names=False)
+                result = bind_cols(res.iloc[:, :-2])
+                tm.assert_series_equal(res['4th'], result, check_names=False)
+                assert result.name is None
+
+                if sort:
+                    tm.assert_frame_equal(
+                        res, res.sort_values(icols, kind='mergesort'))
+
+                out = merge(left, right.reset_index(), on=icols,
+                            sort=sort, how='left')
+
+                res.index = np.arange(len(res))
+                tm.assert_frame_equal(out, res)
+
+        lc = list(map(chr, np.arange(ord('a'), ord('z') + 1)))
+        left = DataFrame(np.random.choice(lc, (5000, 2)),
+                         columns=['1st', '3rd'])
+        left.insert(1, '2nd', np.random.randint(0, 1000, len(left)))
+
+        i = np.random.permutation(len(left))
+        right = left.iloc[i].copy()
+
+        left['4th'] = bind_cols(left)
+        right['5th'] = - bind_cols(right)
+        right.set_index(icols, inplace=True)
+
+        run_asserts(left, right)
+
+        # inject some nulls
+        left.loc[1::23, '1st'] = np.nan
+        left.loc[2::37, '2nd'] = np.nan
+        left.loc[3::43, '3rd'] = np.nan
+        left['4th'] = bind_cols(left)
+
+        i = np.random.permutation(len(left))
+        right = left.iloc[i, :-1]
+        right['5th'] = - bind_cols(right)
+        right.set_index(icols, inplace=True)
+
+        run_asserts(left, right)
+
+    def test_merge_right_vs_left(self, left, right):
+        # compare left vs right merge with multikey
+        for sort in [False, True]:
+            merged1 = left.merge(right, left_on=['key1', 'key2'],
+                                 right_index=True, how='left', sort=sort)
+
+            merged2 = right.merge(left, right_on=['key1', 'key2'],
+                                  left_index=True, how='right',
+                                  sort=sort)
+
+            merged2 = merged2.loc[:, merged1.columns]
+            tm.assert_frame_equal(merged1, merged2)
+
+    def test_compress_group_combinations(self):
```

Comment:
What is this actually testing?

Reply:
I think we try to test a left.join(right) and a symmetric right.join(left)